nomad/plan_apply.go

```diff
@@ -99,7 +100,16 @@ func (p *planner) planApply() {
 		// Snapshot the state so that we have a consistent view of the world
 		// if no snapshot is available
 		if waitCh == nil || snap == nil {
-			snap, err = p.fsm.State().Snapshot()
+			const timeout = 5 * time.Second
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			snap, err = p.fsm.State().SnapshotAfter(ctx, pending.plan.SnapshotIndex)
```

Comment:
We probably should comment why we wait and why it's the `plan.SnapshotIndex` that's worth waiting for.

As a minor nit-pick, I find the name `SnapshotAfter` a bit confusing, it's more `SnapshotWhenAtleast` or `AtOrAfter` rather than `After.

Also, should we handle the case where `snap != nil` but is behind `pending.plan.SnapshotIndex`?  Though, I don't quite follow all of the interactions of loop/waitCh/snap manipulation to fully understand if it's needed.

Reply:
> Also, should we handle the case where snap != nil but is behind pending.plan.SnapshotIndex? Though, I don't quite follow all of the interactions of loop/waitCh/snap manipulation to fully understand if it's needed.

Yup! Definitely a bug.

Reply:
Renamed SnapshotMinIndex with rationale in commit bcad687 message :

> SnapshotMinIndex was chosen as it describes the action (snapshot), a
constraint (minimum), and the object of the constraint (index).